### ISSUE TITLE
Spack is building a non-optimized CMake?

### ISSUE THREAD

**Issue by bartlettroscoe** [spack/spack#14408]
It would seem that the Spack cmake/package.py file is set up to build a **no optimized** version of CMake by default.  I noticed this a few months ago when we were almost going to try to use the Spack-installed env but then noticed this that the Spack-installed cmake was about 2.5x slower than correctly built CMake.  Here are the configure times for the Primary Tested Trilinos suite on my super fast Linux machine:

Install Method | CMake version | Runtime
-- | -- | --
SEMS | 3.12.2 | 2m30.498s
TriBITS | 3.11.1 | 2m29.507s
Spack | 3.14.5 | 7m47.042s

Until we can fix this, CMake as installed by spack is unusable for SNL ATDM :-(

The proper way to build an optimized CMake is shown in:

* https://github.com/TriBITSPub/TriBITS/blob/master/tribits/devtools_install/install-cmake.py#L172

and that shows how to do a fast parallel configure too.

**Comment by chuckatkins** [comment]
While the default build does seem to be unoptimized, just adding `O3` to the flags as the referenced implementation does is not universally safe across all the compliers we need to bootstrap with.  However, we should be able to make sure a release build gets built, which in most cases will use `O2` and should address the performance issues as well as be a universally safe.  I'll get a pr up today for it.

**Comment by bartlettroscoe** [comment]
@chuckatkins, that would be great!   Please let me know when you have a commit that you think works and I will cherry-pick it and test it out on our side.

@robertmaynard, @bradking, what is the best way to build and install CMake on linux systems?  The instructions at:

* https://cmake.org/install/

say to use "bootstrap" but at SNL we have been using `configure` with `--parallel` as shown at:

* https://github.com/TriBITSPub/TriBITS/blob/master/tribits/devtools_install/install-cmake.py#L172

which seems to build a very nicely optimized version of CMake.

**Comment by chuckatkins** [comment]
> Please let me know when you have a commit that you think works and I will cherry-pick it and test it out on our side.

See #14414 

> The instructions say to use "bootstrap" but at SNL we have been using `configure`

`configure` is just a convenience for people familiar with autotools.  If you look at the contents of the `configure` script all it does is call the `bootstrap` script forwarding along all the same arguments.

> which seems to build a very nicely optimized version of CMake.

The "optimized" version you're getting is coming from explicitly setting `-O3` in the `CFLAGS` and `CXXFLAGS`, not from any way you're calling the configure script.  The referenced PR instead achieves this effed by passing `-DCMAKE_BUILD_TYPE=Release` to the cmake build itself.  This will often resolve to `O3` but not always.  It relies on the "optimized" build flags known by CMake to be generally safe.  in addition it also makes sure things like `-DNDEBUG` get set.

**Comment by bartlettroscoe** [comment]
Just an FYI, I cherry-picked that commit to our spack 'snl-atdm' branch and rebuilt and installed CMake 3.14.5. I then ran new configure timings for the entire Primary Tested Trilinos test suite comparing the SEMS-build cmake to the updated Spack-built cmake. I ran the configure 4 times each and the averages were as shown below.

Install Method | CMake version | Runtime
-- | -- | --
SEMS | 3.12.2 | 3m8s
Spack | 3.14.5 | 3m18s

I would say that the Spack-built cmake 3.14.5 is statistically slower than the SEMS-built cmake 3.12.2 configure. But that is still only a 5% increase in overall configure times so I would guess that is due to CMake 3.14.5 being slower than 3.12.2, not due to a problem in how CMake is built in Spack. That is not serious enough to warrant doing any debugging I don't think.

@chuckatkins, thanks for fixing this so quickly!
